File: 10917/main.c
#include <stdio.h>
#include <memory.h>

#define NODES    1001
#define INFINITE 1000001

int path[NODES][NODES];
int route[NODES];

int dfs(int src, int n)
{
   int i;

   if (route[src] == 0) {
      for (i = 1; i <= n; i++) {
         if (path[src][i] != INFINITE && path[2][i] < path[2][src]) {
            route[src] += dfs(i, n);
         }
      }
   }
   return route[src];
}

void shortest_path(int src, int n)
{
   int dest, i, k, d;
   char selected[NODES];

   for (i = 1; i <= n; i++) {
      selected[i] = (i == src) ? 1 : 0;
   }

   for (i = 1; i <= n; i++) {
      k = 0;
      d = INFINITE;
      for (dest = 1; dest <= n; dest++) {
         if (path[src][dest] < d && !selected[dest]) {
            d = path[src][dest];
            k = dest;
         }
      }
      selected[k] = 1;
      for (dest = 1; dest <= n; dest++) {
         d = path[src][k] + path[k][dest];
         if (d < path[src][dest]) {
            path[src][dest] = d;
         }
      }
   }
}

int main(int argc, char* argv[])
{
   int n, m, i, j, k;

   for (i = 0; i < NODES; i++) {
      path[0][i] = path[i][0] = INFINITE;
   }

   while (scanf("%d%d", &n, &m) == 2) {
      for (i = 1; i <= n; i++) {
         for (j = 1; j <= n; j++) {
            path[i][j] = (i == j) ? 0 : INFINITE;
         }
      }
      while (m-- > 0) {
         scanf("%d%d%d", &i, &j, &k);
         path[j][i] = path[i][j] = k;
      }
      shortest_path(2, n);
      memset(route, 0, sizeof(int) * (n + 1));
      route[2] = 1;
      printf("%d\n", dfs(1, n));
   }

   return 0;
}

File: 10917/Main.java
import java.io.*;
import java.util.*;

public class Main {
   final static int INFINITE = 1000001;
   public static void main(String args[]) throws IOException {
      String line;
      StringBuilder answer = new StringBuilder();
      BufferedReader in = new BufferedReader(new InputStreamReader(System.in));

      while ((line = in.readLine()) != null) {
         if (line.charAt(0) == '0' && line.length() == 1)
            break;
         StringTokenizer tokenizer = new StringTokenizer(line);
         int n = Integer.parseInt(tokenizer.nextToken());
         int m = Integer.parseInt(tokenizer.nextToken());
         int[] route = new int[n + 1];
         int[][] matrix = new int[n + 1][n + 1];

         for (int i = 0; i < matrix.length; i++) {
            for (int j = 0; j < matrix.length; j++) {
               matrix[i][j] = (i == j) ? 0 : INFINITE;
            }
         }

         while (m-- > 0) {
            tokenizer = new StringTokenizer(in.readLine());
            int a = Integer.parseInt(tokenizer.nextToken());
            int b = Integer.parseInt(tokenizer.nextToken());
            int d = Integer.parseInt(tokenizer.nextToken());
            matrix[a][b] = matrix[b][a] = d;
         }

         findShortestPath(matrix, 2);
         route[2] = 1;
         answer.append(Integer.toString(dfs(matrix, route, 1)));
         answer.append('\n');
      }
      System.out.print(answer);
   }

   public static int dfs(int[][] matrix, int[] route, int src) {
      if (route[src] == 0) {
         for (int i = 1; i < matrix.length; i++) {
            if (matrix[src][i] != INFINITE && matrix[2][i] < matrix[2][src]) {
               route[src] += dfs(matrix, route, i);
            }
         }
      }
      return route[src];
   }

   public static void findShortestPath(int[][] matrix, int src) {
      boolean[] selected = new boolean[matrix.length];

      selected[src] = true;
      for (int i = 1; i < matrix.length; i++) {
         int k = 0;
         int d = INFINITE;
         for (int dest = 1; dest < matrix.length; dest++) {
            if (matrix[src][dest] < d && !selected[dest]) {
               d = matrix[src][dest];
               k = dest;
            }
         }
         selected[k] = true;
         for (int dest = 1; dest < matrix.length; dest++) {
            d = matrix[src][k] + matrix[k][dest];
            if (d < matrix[src][dest]) {
               matrix[src][dest] = d;
            }
         }
      }
   }
}
